Auth feature

jfranciscosousa opened the pull request:
Placeholder PR for the meantime

jfranciscosousa committed:
Add password and email to firefighters

Adds a password hashed with `bcrypt`.
Also adds routes to create and auth users.
Currently the authentication just returns the user if the password is correct, I'll add JWT tokens later.

jfranciscosousa committed:
Add JWT based authentication

Secure all application routes with a JWT middleware. Sign the `email` on the token and give it an expiry time of one hour.
Also uses `dotenv` to load env vars.
The JWT signing and verification uses the KEY specified on the `.env` file.

jfranciscosousa committed:
Setup rosie, a factory girl clone

Also fixes some dependencies location (dev vs regular). Removes cluster, useless dependencies.

luistorres commented on src/backend/test/support/factories.js:
```diff
@@ -0,0 +1,19 @@
+  name: faker.name.firstName,
+  email: faker.internet.email(),
+  password: "foobar",
+  status: sample("active", "inactive", "busy")
```
We are missing one attr that is `dutyType`

jfranciscosousa replied:
Not need for the tests though, wanna add it?

jfranciscosousa replied:
Besides, dont write reviews here, this is a long standing PR :stuck_out_tongue:

luistorres replied:
I know, but had nowhere else to comment. And duty is not working in the app :/

jfranciscosousa committed:
Change email to code

jfranciscosousa commented:
@luistorres anything else you need?

luistorres commented:
Maybe assuring that the `dutyType` is working as I don't see it anywhere besides the seeds.
Besides that, do we have endpoints for changing password and login out? If so, all is good I think

jfranciscosousa commented:
We don't have a logout route, we are using stateless tokens so the only possibility would be to implement a token blacklist. As I told you before the best possible way would be to have short-lived tokens (30 mins or so) and to refresh the token on the client-side every 30 mins.

jfranciscosousa commented:
About the `dutyType` I really can't understand the problem, I didn't meddle with it at all.

luistorres commented:
30 mins is really low. I would say at least 7 days or so, and every visit to the app in FE could try to refresh it (for instance). Token expiration in this app is just a pain, so the more we can delay it, the better.
Logout is needed just in case you want to change user...we shouldn't just clean the token in FE, that is not the best practice,  though for now it is enough.

There is no problem with `dutyType`, I'm just saying we can have tests, as we have for all the other things.

jfranciscosousa commented:
To the first question: I'll add session-based auth then.

To the other question: I'll add the duty type then.

jfranciscosousa commented:
Forget the session-based auth on node.js, I can't do it (no time for it).

About cleaning the token... JWT is a stateless authentication method, there is no "logout" and it's used for simple APIs. It's easier to test and develop with on Node, rather than using cookies and sessions and whatnot. And even with session and cookie-based mechanisms, you have the same problems that you have with stateless, in that you have no way of securing your app if a hacker manages to capture a token on requests.

If you use HTTPS 99.99% no problems with people getting the token (unless they hack HTTPS on itself). The only sure way people can hack something is if they open the user's browser and fetch the token from localStorage (they would need to actually steal a computer or a phone). Native apps use SecureStorage APIs on Android and iOS, but then again physical access means you are f****.

On a breach, you can change the `JWT_SIGN_KEY` variable and all of the tokens are invalidated.

As of now and the future. 7 days of the expiry time. Every time the user opens the app you refresh the token. Check out the spec of OAuth. They work by having a short-lived token and a long-lived refresh token (the second one lives on the browser). You use the short-lived token in exchanges with a server and then you use the refresh token to ask for a new token every now and then. 

Currently, I don't plan on implementing refresh tokens, I would just create a route that returns a newly built JWT. These things are hard to make and test on part-time. Perhaps moving to Rails would be a good idea but now it's too late.

If this would be a native app, you could just store the token in SecureSTorage and use biometrics to fetch it.

jfranciscosousa commented:
Things to do:
* Add refresh token route
* Add dutyType to tests
* Add role-based authorization (admin vs regular user)
* Add authorization to the `socket.io` module (should be admin only)

luistorres commented:
I'm good with everything. If we are able to change `JWT_SIGN_KEY ` and invalidate everything, can we have the expiration be something like 3 months then? Even 7 days seem too little when guys are expected to open this like once a month or something.

luistorres commented:
We need 3 degrees of authorization actually:
- user only gets to new availability screen, so basically can only change it's own status.
- dashboard (only sees dashboard) - can manage everyone status and dutyType
- admin - sees dashboard (change status and dutyType), can add/remove users, can send notifications (to be done yet)

jfranciscosousa committed:
Remove the auth requirement from the login step

jfranciscosousa committed:
Add authentication to socket

https://socket.io/docs/client-api/#With-extraHeaders use this to add headers on the client.

Also has example error handler on the client.

jfranciscosousa commented:
@luistorres given new developments, can I close this and also archive the repo?
